// @what3words/javascript-components/dist/collection/lib/sdk.js
import what3words, { ApiVersion, AutosuggestClient, AvailableLanguagesClient, ConvertToCoordinatesClient, ConvertTo3waClient, GridSectionClient, fetchTransport, } from '@what3words/api';
import { W3W_REGEX } from './constants';
const api = what3words('', { apiVersion: ApiVersion.Version3 }, { transport: fetchTransport() });
const utils = {
  validAddress: (v) => W3W_REGEX.test(v),
};
const initClient = (clientName, client) => {
  return (apiKey, config, transport) => {
    let clientApiKey;
    if (apiKey) {
      clientApiKey = apiKey;
    }
    else {
      console.warn('Use of sdk script credentials is deprecated! Please pass the `api_key` attribute to components. https://developer.what3words.com/tutorial/javascript-autosuggest-component-v4');
      clientApiKey = api.clients[clientName].apiKey();
    }
    return new client(clientApiKey, config, transport);
  };
};
export const what3wordsClients = {
  autosuggest: initClient('autosuggest', AutosuggestClient),
  availableLanguages: initClient('availableLanguages', AvailableLanguagesClient),
  convertTo3wa: initClient('convertTo3wa', ConvertTo3waClient),
  convertToCoordinates: initClient('convertToCoordinates', ConvertToCoordinatesClient),
  gridSection: initClient('gridSection', GridSectionClient),
};
export const sdk = {
  api,
  utils,
};
