// @what3words/javascript-components/dist/collection/components/address/address.js
import { h } from '@stencil/core';
import { DEFAULTS } from '../../lib';
export class What3wordsAddress {
  constructor() {
    this.words = undefined;
    this.iconColor = DEFAULTS.iconColor;
    this.textColor = DEFAULTS.textColor;
    this.size = DEFAULTS.addressSize;
    this.target = DEFAULTS.target;
    this.link = DEFAULTS.true;
    this.tooltip = DEFAULTS.true;
    this.tooltipLocation = DEFAULTS.tooltipLocation;
    this.showTooltip = DEFAULTS.false;
  }
  render() {
    const url = `https://map.what3words.com/${this.words}`;
    const attributes = this.link ? { href: url, target: this.target } : {};
    return (h("a", Object.assign({}, attributes, { class: `what3words-address notranslate ${this.showTooltip ? 'what3words-address_tooltip' : ''}`, style: { fontSize: `${this.size}px` } }), this.tooltip && (h("div", { class: "what3words-tooltip-container" }, h("div", { class: "what3words-tooltip" }, "what3words gives every 3m x 3m in the world a unique 3 word address. This one describes the precise ", this.tooltipLocation, ".", ' ', h("a", { href: url, target: this.target }, "View on map")))), h("div", { class: "what3words-address_container" }, h("what3words-symbol", { size: this.size * 1.25, color: this.iconColor }), h("span", { class: "what3words-address_text", style: { color: this.textColor } }, this.words))));
  }
  static get is() { return "what3words-address"; }
  static get originalStyleUrls() {
    return {
      "$": ["address.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["address.css"]
    };
  }
  static get properties() {
    return {
      "words": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "words",
        "reflect": false
      },
      "iconColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "icon-color",
        "reflect": false,
        "defaultValue": "DEFAULTS.iconColor"
      },
      "textColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "text-color",
        "reflect": false,
        "defaultValue": "DEFAULTS.textColor"
      },
      "size": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "DEFAULTS.addressSize"
      },
      "target": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "target",
        "reflect": false,
        "defaultValue": "DEFAULTS.target"
      },
      "link": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "link",
        "reflect": false,
        "defaultValue": "DEFAULTS.true"
      },
      "tooltip": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "tooltip",
        "reflect": false,
        "defaultValue": "DEFAULTS.true"
      },
      "tooltipLocation": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "tooltip-location",
        "reflect": false,
        "defaultValue": "DEFAULTS.tooltipLocation"
      },
      "showTooltip": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "show-tooltip",
        "reflect": false,
        "defaultValue": "DEFAULTS.false"
      }
    };
  }
}
