// @what3words/javascript-components/dist/cjs/what3words-address.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-c574deb6.js');
const index$1 = require('./index-c80d121f.js');

const addressCss = "what3words-address{display:inline-block;font-family:\"Source Sans Pro\", sans-serif}.what3words-address{color:#333;display:-ms-inline-flexbox;display:inline-flex;text-decoration:none;-ms-flex-direction:column;flex-direction:column}.what3words-address_container{display:-ms-flexbox;display:flex}.what3words-address_text{white-space:nowrap}.what3words-tooltip-container{height:0;width:0;position:relative;-ms-flex-item-align:center;align-self:center}.what3words-tooltip{position:absolute;display:none;width:160px;background-color:#333333;border-radius:3px;bottom:10px;left:-88px;font-size:12px;padding:8px;z-index:1;color:#ffffff;-webkit-box-shadow:0 0 16px rgba(0, 0, 0, 0.5);box-shadow:0 0 16px rgba(0, 0, 0, 0.5)}.what3words-address:hover .what3words-tooltip,.what3words-address_tooltip .what3words-tooltip{display:block}.what3words-tooltip:after{bottom:-13px;left:50%;content:\"\";height:0;width:0;position:absolute;border:7px solid transparent;margin-left:-7px;border-top-color:#333333;z-index:1}.what3words-tooltip:before{content:\"\";position:absolute;width:100%;height:10px;bottom:-10px;left:0}.what3words-tooltip a{color:#ffffff}";

const What3wordsAddress = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.words = undefined;
    this.iconColor = index$1.DEFAULTS.iconColor;
    this.textColor = index$1.DEFAULTS.textColor;
    this.size = index$1.DEFAULTS.addressSize;
    this.target = index$1.DEFAULTS.target;
    this.link = index$1.DEFAULTS.true;
    this.tooltip = index$1.DEFAULTS.true;
    this.tooltipLocation = index$1.DEFAULTS.tooltipLocation;
    this.showTooltip = index$1.DEFAULTS.false;
  }
  render() {
    const url = `https://map.what3words.com/${this.words}`;
    const attributes = this.link ? { href: url, target: this.target } : {};
    return (index.h("a", Object.assign({}, attributes, { class: `what3words-address notranslate ${this.showTooltip ? 'what3words-address_tooltip' : ''}`, style: { fontSize: `${this.size}px` } }), this.tooltip && (index.h("div", { class: "what3words-tooltip-container" }, index.h("div", { class: "what3words-tooltip" }, "what3words gives every 3m x 3m in the world a unique 3 word address. This one describes the precise ", this.tooltipLocation, ".", ' ', index.h("a", { href: url, target: this.target }, "View on map")))), index.h("div", { class: "what3words-address_container" }, index.h("what3words-symbol", { size: this.size * 1.25, color: this.iconColor }), index.h("span", { class: "what3words-address_text", style: { color: this.textColor } }, this.words))));
  }
};
What3wordsAddress.style = addressCss;

exports.what3words_address = What3wordsAddress;
