// @what3words/javascript-components/dist/collection/components/symbol/symbol.js
/* eslint-disable @typescript-eslint/no-unused-vars */
import { h, Host } from '@stencil/core';
import { DEFAULTS } from '../../lib';
export class What3wordsSymbol {
  constructor() {
    this.color = DEFAULTS.iconColor;
    this.size = DEFAULTS.symbolSize;
  }
  render() {
    return (h(Host, null, h("svg", { viewBox: "0 0 32 32", class: "what3words-logo", style: {
        color: this.color,
        width: `${this.size}px`,
        height: `${this.size}px`,
      }, "data-testid": "what3words-symbol" }, h("path", { fill: "currentColor", d: "M10.7,4h2L4,28H2L10.7,4z M19.7,4h2L13,28h-2L19.7,4z M28.7,4h2L22,28h-2L28.7,4z" }))));
  }
  static get is() { return "what3words-symbol"; }
  static get originalStyleUrls() {
    return {
      "$": ["symbol.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["symbol.css"]
    };
  }
  static get properties() {
    return {
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "color",
        "reflect": false,
        "defaultValue": "DEFAULTS.iconColor"
      },
      "size": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "DEFAULTS.symbolSize"
      }
    };
  }
}
